rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+  default deployment configuration.
+  The goal is to eliminate any non-required permissions - we don't have to keep
+  anything extra there for demonstration purposes.
+
+  We also have to document all possible required permissions, so that users are
+  aware of the possible configuration options. At Helm Charts we should allow
+  configuring arbitrary permissions via values (while providing sane defaults).
+
+#### Securing secrets
+
+Vector sometimes needs access to secrets, like AWS API access tokens and so on.
+That data has to be adequately protected.
+
+We should recommend users to use [`Secret`][k8s_api_secret]
+([docs][k8s_docs_secret]) instead of [`ConfigMap`][k8s_api_config_map] if they
+have secret data embedded in their Vector `.toml` config files.
+
+We should also consider integrating with tools like [Vault] and [redoctober].
+
+#### Recommend users additional steps to secure the cluster
+
+- Suggest using [Falco].
+- Suggest setting up proper RBAC rules for cluster operators and users;
+  [`audit2rbac`](https://github.com/liggitt/audit2rbac) is a useful tool to
+  help with this.
+- Suggest using [Pod Security Policies][k8s_docs_pod_security_policiy]
+  ([API][k8s_api_pod_security_policy]).
+- Suggest using [NetworkPolicy][k8s_api_network_policy].
+- Suggest runnig [kube-bench].
+- Suggest reading the
+  [Kubernetes security documentation][k8s_docs_securing_a_cluster].
+
+#### Automatic container rebuilds
+
+The ability to rebuild containers with a CVE fix automatically quickly is a very
+important part of a successful vulnerability mitigation strategy.
+We should prepare in advance and rollout the infrastructure and automation to
+make it possible to rebuild the containers for _all_ (not just the latest or
+nightly!) the supported Vector versions.
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+1. [Fluentd k8s integration]
+1. [LogDNA k8s integration]
+1. [Honeycomb integration]
+1. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+1. [Influx Helm charts]
+1. [Awesome Operators List] - an "awesome list" of operators.
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community-driven
+   integrations.
+
+## Outstanding Questions
+
+### From Ben
+
+1. ~~What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].~~
+   See the [Minimal supported Kubernetes version][anchor_minimal_supported_kubernetes_version]
+   section.
+1. ~~What is the best to avoid Vector from ingesting it's own logs? I'm assuming
```

Comment:
Maybe I missed this, but where is in the RFC is this discussed? I assume we're excluding the entire `vector` namespace, right?

Reply:
My plan was to implement this via metadata filtering. I.e. we'll annotate Vector `Pod`s with `vector.dev/exclude: true`, and I'll naturally filter itself out.

Reply:
Got it. Yes, a section would be nice. And I assume this is fine performance-wise?

Reply:
Well, it is extra work, but it doesn't seem like it's going to be a bottleneck for any performance-sensitive paths of Vector.
This is pretty much what the [Filtering based on Kubernetes API metadata](https://github.com/timberio/vector/blob/k8s-rfc/rfcs/2020-04-04-2221-kubernetes-integration.md#filtering-based-on-kubernetes-api-metadata) section is about.

The problem is without this we'll be lacking some key features. So the tradeoff there is to do a bit of extra work and add some (a lot of?) code to implement the best-in-class feature set.

Btw, I added a section on namespace annotation-based filtering there.